Sources/OpenCombine/Publishers/Publishers.Filter.swift

```diff
@@ -0,0 +1,213 @@
+            upstream.receive(subscriber: filter)
+        }
+    }
+}
+
+extension Publisher {
+
+    /// Republishes all elements that match a provided closure.
+    ///
+    /// - Parameter isIncluded: A closure that takes one element and returns
+    /// a Boolean value indicating whether to republish the element.
+    /// - Returns: A publisher that republishes all elements that satisfy the closure.
+    public func filter(
+        _ isIncluded: @escaping (Self.Output) -> Bool
+    ) -> Publishers.Filter<Self> {
+        return Publishers.Filter(upstream: self, isIncluded: isIncluded)
+    }
+
+    /// Republishes all elements that match a provided error-throwing closure.
+    ///
+    /// If the `isIncluded` closure throws an error, the publisher fails with that error.
+    ///
+    /// - Parameter isIncluded:  A closure that takes one element and returns a
+    /// Boolean value indicating whether to republish the element.
+    /// - Returns:  A publisher that republishes all elements that satisfy the closure.
+    public func tryFilter(
+        _ isIncluded: @escaping (Self.Output) throws -> Bool
+    ) -> Publishers.TryFilter<Self> {
+        return Publishers.TryFilter(upstream: self, isIncluded: isIncluded)
+    }
+}
+
+private final class _Filter<Upstream: Publisher, Downstream: Subscriber>
+    : _BaseFilter<Upstream, Downstream>,
+      Subscriber,
+      CustomStringConvertible
+    where Upstream.Output == Downstream.Input,
+          Upstream.Failure == Downstream.Failure {
+
+    var description: String { return "Filter" }
+
+    init(downstream: Downstream, isIncluded: @escaping (Input) -> Bool) {
+        super.init(downstream: downstream,
+                   errorTransform: { $0 },
+                   isIncluded: { .success(isIncluded($0))
+        })
+    }
+}
+
+private final class _TryFilter<Upstream: Publisher, Downstream: Subscriber>
+    : _BaseFilter<Upstream, Downstream>,
+    Subscriber,
+    CustomStringConvertible
+    where Upstream.Output == Downstream.Input, Downstream.Failure == Error {
+
+    init(downstream: Downstream, isIncluded: @escaping (Input) throws -> Bool) {
+        super.init(downstream: downstream,
+                   errorTransform: { $0 as Error },
+                   isIncluded: { input in Result { try isIncluded(input) }
+        })
+    }
+
+    var description: String { return "TryFilter" }
+}
+
+private class _BaseFilter<Upstream: Publisher, Downstream: Subscriber>
+    : OperatorSubscription<Downstream>,
+      Subscription
+    where Upstream.Output == Downstream.Input
+{
+    typealias Input = Upstream.Output
+    typealias Output = Downstream.Input
+    typealias Failure = Upstream.Failure
+
+    private let _isIncluded: (Input) -> Result<Bool, Downstream.Failure>
+    private var _unsatisfiedDemand: Subscribers.Demand = .none
+    private var _isFinished = false
+
+    var errorTransform: (Upstream.Failure) -> Downstream.Failure
```

Comment:
Why is this property needed?

Reply:
This allows use to use this in `receive` so that we don't have to write out a different "receive" for every subclass that is using _Filter and instead maps the error.  Otherwise we have an issues with generic typing that you need to force cast the Failure as! Downstream.Failure.